### ISSUE TITLE
(Mozilla 1437080) service workers: load after unload, or session restoration, may lead to 'Corrupted Content Error' or 'Cannot load app'

### ISSUE THREAD

**Issue by grahamperrin** [BrowserWorks/waterfox#581]
## Steps

1. https://meta.discourse.org/t/-/78008/24?u=grahamperrin
2. open a new tab
3. quit
4. start
5. restore the previous session
6. bring the Discourse tab to the front

## Expected

7. automated load of the content

## Actual result

7. no load; 

> # Corrupted Content Error
> 
> The site at `https://meta.discourse.org/t/cannot-load-app-for-a-meta-topic/78008/24` has experienced a network protocol violation that cannot be repaired.
> 
> The page you are trying to view cannot be shown because an error in the data transmission was detected.
> 
> - Please contact the website owners to inform them of this problem.
> 
> **Try Again**

## Workaround

<kbd>Space</kbd>
 
Alternative workarounds include <kbd>Control</kbd>-<kbd>L</kbd> then <kbd>Return</kbd> or <kbd>Enter</kbd>, however for many users, it will be preferable to simply accept the invitation: 

- _try again_. 

----

<details>
<summary>
Notes
</summary>
If you see 'Cannot load app' in lieu of 'Corrupted Content Error', please state the build of Waterfox, from: 

- about:support

Please refrain from contacting the owners of meta.discourse.org. For this issue, re: http://www.discourse.org/about#browser we may expect the current release of Waterfox to be not supported by Discourse. 

If you find an instance of Discourse that is not affected by this issue, please state both: 

- the URL; and 
- the version of Discourse.

Version information can be shown with [Wappalyzer](https://addons.mozilla.org/addon/wappalyzer/) or [its bookmarklet](https://www.wappalyzer.com/download#bookmarklet).

Thanks
</details>

**Comment by grahamperrin** [comment]
Meta, tracking: #538 

Reproducible with Waterfox 56.1.0 on a Mac. 

<details>
<summary>
Additional information
</summary>
Curiously, I began noticing this issue around the time of testing and release of 56.2.0. For example: 

![2018-05-20 19 53 35 version 2](https://user-images.githubusercontent.com/192271/40468831-70e91756-5f26-11e8-80f2-676ac3079bbb.png)

> Cannot load app

That was at home, https://discourse.mozilla.org/t/mozilla-buys-pocket-ask-your-questions-here/13998 with <strike>Discourse 2.0.0 </strike>I don't know which version of Discourse, and _almost_ certainly version 56.2.0.7_1 of Waterfox on FreeBSD-CURRENT. 

Today at home for the same URL, without a quit, I get the alternative symptom: 

> Corrupted Content Error

Discourse 2.0.0. 

```text
$ pkg info waterfox | grep Version
Version        : 56.2.0.13_1
$ date ; uname -v
Thu 24 May 2018 07:51:14 BST
FreeBSD 12.0-CURRENT #2 r333587: Mon May 14 01:59:02 BST 2018     root@momh167-gjp4-hpelitebook8570p-freebsd:/usr/obj/usr/src/amd64.amd64/sys/GENERIC 
```

– 56.2.0 build 20180523041847
</details>

**Comment by grahamperrin** [comment]
> Reproducible with Waterfox 56.1.0 on a Mac.

## Refinement

- 55.1.0: not reproducible
- 56.0.3: not reproducible
- 56.1.0 with Stylo disabled: reproducible. 

I might find a Mozilla bug for this …

**Comment by grahamperrin** [comment]
Symptoms of probably the same bug (I might need to change the title here): 

1. https://developer.microsoft.com/en-us/microsoft-edge/ not signed in
2. click **Sign in**
3. do not proceed with the sign in routine
4. back

### Result

5. _Corrupted Content Error …_ at https://developer.microsoft.com/en-us/microsoft-edge/

**Comment by grahamperrin** [comment]
Mozilla bug [1454400 - Session Restore doesn't activate service worker](https://bugzilla.mozilla.org/show_bug.cgi?id=1454400) was a duplicate of [1437080 - service workers gets wrong FetchEvent.request.cache value](https://bugzilla.mozilla.org/show_bug.cgi?id=1437080)

>  RESOLVED FIXED in Firefox 60

- https://bugzilla.mozilla.org/show_bug.cgi?id=1437080#module-attachments

– and from <https://bugzilla.mozilla.org/show_bug.cgi?id=1437080#c15>, 

- https://hg.mozilla.org/mozilla-central/rev/3e216f67b25f
- https://hg.mozilla.org/mozilla-central/rev/4a3044a631d7
- https://hg.mozilla.org/mozilla-central/rev/784e9f86bb03

**Comment by grahamperrin** [comment]
Not all service worker-active PWAs are affected. 

The _Video Trimmer_ application at https://paulkinlan.github.io/videocut/index.html loads without error after a manual unload.

**Comment by L-a-n-g-o-l-i-e-r-s** [comment]
dom.serviceWorkers.enabled true for me results in "ghost window" from service workers related to outlook and youtube, they don't close/stop after the related tabs have been closed and remain open until the browser is entirely closed.

I noticed that when this happened my browser would progressively become more sluggish, experience lock ups and eventually the memory usage would climb to 1-2 gb. When I toggled dom.serviceWorkers.enabled to false, this behavior stopped entirely and WF was back to its usual snappiness. 

For the record I am just looking for commiseration, not to submit an in depth report for lack of time. :disappointed:

**Comment by MrAlex94** [comment]
https://github.com/MrAlex94/Waterfox/issues/694#issuecomment-404564400

Could you see if these new commits fix it? I can't seem to repro so far :-0

**Comment by grahamperrin** [comment]
Thanks. 

56.2.1.55_3 at https://www.freshports.org/www/waterfox is a fraction behind, and I don't have a Windows test environment handy, so I might be unable to verify until Monday or later. Or … 

… _if_ you have a `.dmg` of the pre-release handy, and if you can get it to me this morning (Friday) – PM or chat in Reddit, I guess – then I might be able to verify before the end of this afternoon. 

In any case, don't let my lack of verification delay a release. If you made the issue reproducible, yourself, _before_ the commits: that's good enough for me :+1: 

----

Bottom line: a fix will be smart but IMHO this issue is not a showstopper. If I understand correctly, the _reported_ corruption is cache-specific; not a cause for concern (e.g. no data loss).

**Comment by grahamperrin** [comment]
From <https://github.com/MrAlex94/Waterfox/issues/581#issuecomment-391608566>: 

> Reproducible with Waterfox 56.1.0 on a Mac.

Unfortunately I no longer have the affected profile. Either that, or the service worker was automatically un-registered at some point. 

With 56.2.1 (and 56.2.2) I can 'fake' registration by editing then locking `serviceworker.txt`, but I shouldn't treat this fakery as a reliable step towards verifying a fix. The worker remains active after registration disappears (from `about:serviceworkers`) and registration will reappear only after a quit of Waterfox, with the lock on the file … and so on. Feels clumsy. 

With build 20180714033244 of 56.2.2 I can reproduce the symptom if there's no network connection: 

![2018-07-14 21 50 34](https://user-images.githubusercontent.com/192271/42728498-70939672-87b3-11e8-8c39-ea7198b46d79.png)

![2018-07-14 21 51 16](https://user-images.githubusercontent.com/192271/42728500-780f3b68-87b3-11e8-8f4d-0348eacba6d7.png)

– however, that's an edge case (I did **not** make any offline test in May or June) and maybe a separate Mozilla bug. I shouldn't keep this issue open for that alone. 

----

For me it'll be sane to verify in a FreeBSD environment where I need not fake service worker registration. However we should not assume that the www/waterfox experiment will reach 56.2.2. 

More practically: I reckon, seek verification of the fix in Windows case #694 after 56.2.2 is released.

**Comment by grahamperrin** [comment]
Thanks @jbeich for making 56.2.2 available for FreeBSD. 

> … With build 20180714033244 of 56.2.2 (on Mac OS X 10.9.5) I can reproduce the symptom if there's no network connection: …

The same, or similar, with 56.2.2 on FreeBSD-CURRENT. 

I might attach (or link to) a screen recording but essentially, it seems to happen in some cases if: 

- there's no network connection **and**
- the worker for the affected page is not active before an unload

– then a load (after unload) may report corrupted content, and clicking **Try again** will not progress to a more user-friendly _Server not found_ page.

As far as I could tell it seemed to affect e.g. Mozilla Discourse but not CNET (tested: [Magic Leap's long-awaited headset will ship this summer -- and everything else we learned - CNET](https://www.cnet.com/news/magic-leap-long-awaited-headset-will-ship-this-summer-and-everything-else-we-learned/)). 

Let's close this. Thanks to Alex and Jan for the commits.

**Comment by grahamperrin** [comment]
### Two screen recordings

Mixtures of online and offline with 56.2.2 on FreeBSD-CURRENT. 

[2018-07-15 11:29.mp4.zip](https://github.com/MrAlex94/Waterfox/files/2196421/2018-07-15.11.29.mp4.zip)

[2018-07-15 11:56.mp4.zip](https://github.com/MrAlex94/Waterfox/files/2196427/2018-07-15.11.56.mp4.zip)

From the second recording: 

- 04:47 on the timeline
- going offline

![04 47](https://user-images.githubusercontent.com/192271/42741962-a873fa7c-88ae-11e8-8a49-1e0b270068c4.png)

- 06:00 
- offline
- service worker active
- unloading

![06 00](https://user-images.githubusercontent.com/192271/42741713-9ec30d8a-88ac-11e8-958c-ee006f4aea92.png)

- 06:11
- still offline
- loading

![06 11](https://user-images.githubusercontent.com/192271/42741810-6abd720e-88ad-11e8-9647-ee1dafe624ce.png)

- 06:33
- still offline
- service worker not active
- reloading

![06 33](https://user-images.githubusercontent.com/192271/42741791-401152dc-88ad-11e8-9369-894e4f0c908e.png)

- 06:34
- still offline
- three frames – reloaded, blank, corrupted content error

![06 34 a](https://user-images.githubusercontent.com/192271/42741897-240261fc-88ae-11e8-9353-3239e3d218a4.png)

![06 34 b](https://user-images.githubusercontent.com/192271/42741902-2c6fff98-88ae-11e8-8fd0-a22914ed4229.png)

![06 34 c](https://user-images.githubusercontent.com/192271/42741906-31c20e14-88ae-11e8-8d2b-1b05c8eb5301.png)

**Comment by grahamperrin** [comment]
### Comparison with Firefox Quantum

Firefox 61.0.1 (64-bit) in safe mode on Kubuntu 17.10 on an old Ergo Vista 621 notebook. 

https://roadtrip.glitch.me/ around forty-five minutes ago: 

![2018-07-16 03 51 25](https://user-images.githubusercontent.com/192271/42742312-bd1550e0-88b1-11e8-94c3-19fa1cd09cbe.png)

I can not make it reproducible but I guess that there truly **was** a

> network protocol violation

Re: <https://github.com/cramforce/roadtrip/issues/7#issuecomment-404258068> I had previously left the application, in Firefox, [Finding my location](https://screenshots.firefox.com/DVezgwaQtOgUeiJL/roadtrip.glitch.me) then closed the notebook for ... I don't know how long. Days, probably. 

This morning the reported corrupted content error (network protocol violation) occurred very shortly after I opened the notebook. Can't recall whether I attempted to reload the page.

**Comment by reitzig** [comment]
I don't use Waterfox, but this was the best hit on Google. So, for the record, I still experience the error with FF 62.0.3 (Ubuntu 18.04).

**Comment by grahamperrin** [comment]
@reitzig please raise the issue in a support area for Firefox. If you already use Reddit, https://www.reddit.com/r/firefox/ might be ideal. 

I'll join you there (not here; this issue is closed). Thanks.
